### DISCUSSION TITLE
Is Github Copilot safe privacy wise?

### DISCUSSION THREAD

**Discussion by haseeb090** [community/community#7163] [Copilot Conversations] [answered]
Hi 
I'm sure this is a stupid question and someone else probably asked it. But is Github Copilot safe to be used on private projects/repos. I'm sure you collect prediction analytics but that won't entail collecting actual code right?
Like would i need to be worried using copilot in my work? Would it in any way possibly put my organization's codebase or data at risk??
Hope that made sense.
Looks amazing though. Cheers!

**Comment by davecheney** [comment, marked as answer]
The data copilot collects is detailed in the terms of service https://docs.github.com/en/github/copilot/github-copilot-telemetry-terms

**Reply by flavienbwk** [reply]
This should be an option.

**Reply by Temepest74** [reply]
So the answer is maybe. But please add an option to disable it. Or at least make this the default behavior for private repo.

**Reply by leonardBang** [reply]
For this reason, I’ve stopped using Github Copilot.

**Reply by frozenprakash** [reply]
The link is dead.

**Reply by NeuralAIM** [reply]
![image](https://user-images.githubusercontent.com/75082388/175054706-3b67a916-b393-4b2c-975d-86094b40c5ec.png)

**Reply by frozenprakash** [reply]
While signing for GitHub CoPilot, it asks for two checkboxes.
1. To use code from public repository - Check it
2. GitHub copilot will use our code for improvement of it's algorithm - Nope - Unckeck it

Remember to unckeck the 2nd checkbox to avoid exposing your code from any of your repository.

**Reply by antoniohernan** [reply]
Where can this document be obtained?

https://docs.github.com/en/github/copilot/about-github-copilot-telemetry
https://docs.github.com/en/github/copilot/telemetry-terms

It appears as 404, any public version/copy of the document please?

Thanks!

**Reply by frozenprakash** [reply]
@antoniohernan 
GitHub CoPilot Installation:
https://github.blog/2022-06-21-github-copilot-is-generally-available-to-all-developers/

This documentation should get you going right away:
https://github.com/github/copilot-docs

**Comment by khushnoodasif** [comment]
I think the whole future is scary but exciting. Yes Copilot does collect personal data so just take precaution when working in private repos.

**Reply by sadransh** [reply]
just to clarify, it might share the .py/.ipynb/.json file is opened in vscode with Microsoft/OpenAI. Is that correct?

**Reply by luccahuguet** [reply]
> just to clarify, it might share the .py/.ipynb/.json file is opened in vscode with Microsoft/OpenAI. Is that correct?

that's how it seems from what I read of the terms of service

**Reply by moritztim** [reply]
> sadransh

yes but that won't be directly suggested to other users, it will just teach github copilot how to write code. They also have a system in place to prevent specific data from being copied (like tokens or simmilar)

**Comment by MatkoMilic** [comment]
If my code is being collected, arent I putting my companies assets at risk by using Copilot?

**Reply by sadransh** [reply]
so the logic is simple if you accepted the co-pilot telemetry then you already know that you might be sending codes to MSFT/Open AI. 
Yes, you should discuss with your manager regarding using this on private commercial projects. 
so, I personally don't risk using it on private company owned code bases.

**Comment by Tall1n** [comment]
> Is Github Copilot safe privacy wise? #7163

# No.

--- 

### This is the only real answer that should be on top of this post...

Details From the ToS [highlighted by me]:
> when you edit files with the GitHub Copilot extension/plugin enabled, **file content snippets**, suggestions, and any modifications to suggestions **will be shared with GitHub, Microsoft, and OpenAI**, and used for diagnostic purposes **to improve suggestions and related products**. 

They add
> **GitHub Copilot does not use** these URLs, file paths, or **snippets collected** in your telemetry **as suggestions for other users of GitHub Copilot.** 

and 
> This **usage information is used** by GitHub, and shared with Microsoft and OpenAI, **to develop and improve the extension**/plugin **and related products.** 

So they scan all files and content and save it to "improve their product". 

---

Only way to get the data removed according to the ToS:
> You may revoke your consent to the telemetry and personal data processing operations described in this paragraph by contacting GitHub and requesting removal from the technical preview.

So the only way to have your private data removed is: 
Contacting them and telling them to remove you from the beta and requesting them to delete your data collected by copilot.

--- 

Speculation from my side:
If they want to improve their product without using our data as suggestions, they can not use it in the model training data, or the language model would suggest them sooner or later. What they could do is use it for validation/testing of their models output. If you got a "read_from_db.py" file using the "pymysql" package, they can check if their model results are similar to yours when given these two (or similar) prompts.

There should definitly be an option to opt-out, but I assume their idea is to just collect everything and make the best product possible, whatever the privacy implications may be.

**Reply by 8ctopus** [reply]
Best answer.

**Reply by ghost** [reply]
I have a feeling there will be more optionality in terms of privacy if and when the product goes out of beta. Since it's in 'beta' they will probably ask for more collection for diagnostics and improvements - and that is why it is required when enrolled in beta and using it.

However, I did notice in 2 seperate python projects using copilot, it suggests variables and data usage from my other project, which I did not expect, since they are seperated projects..

**Reply by culpritgene** [reply]
Do they collect data only from the same repo, or maybe even from all places accessible to my GitHub account? Like, does it operate on the level of GitHub (my remote repos) or on the level of particular IDE?
Also does copilot work only through the IDE extension or does it analyze my GitHub code in the background even if it is not installed to IDE?
If I will join the copilot program and use it in a separate IDE for personal projects will it still be able to see my private code?

**Reply by moritztim** [reply]
> I have a feeling there will be more optionality in terms of privacy if and when the product goes out of beta. Since it's in 'beta' they will probably ask for more collection for diagnostics and improvements - and that is why it is required when enrolled in beta and using it.
> 
> However, I did notice in 2 seperate python projects using copilot, it suggests variables and data usage from my other project, which I did not expect, since they are seperated projects..

perhaps you have your own local profile where it learns about how you code 🤔  I just hope it stays local...

**Comment by davecheney** [comment]
You may opt out of telemetry collection via your copilot settings.

**Reply by Tall1n** [reply]
Is that now available with the new, paid version? I could not find any information about that.

**Reply by davecheney** [reply]
It is available to all individuals.

**Reply by trompx** [reply]
Opting out doesn't prevent copilot to [send all your data](https://www.reddit.com/r/programming/comments/vhnvlm/comment/idaad4b/?utm_source=share&utm_medium=web2x&context=3) tough

> If I understand the settings correctly then unchecking "Allow GitHub to use my code snippets for product improvements" should disable that.

> No, that's what allows them to use your code. No matter what, all the 'AI' stuff is done on a remote server that receives all your code. In most companies sending source code to a 3rd party without permission will get you fired almost instantly.

**Reply by SkyFoxCoder** [reply]
"If it's free, YOU are the product"
That never failed me.
I know, copilot isn't free, only the first 60 days, but what I mean by that is that they use that 60 days to collect as MUCH data as possible on everyone using/trying it. With that, they are in a much better position business-wise.

**Reply by moritztim** [reply]
> "If it's free, YOU are the product" That never failed me. I know, copilot isn't free, only the first 60 days, but what I mean by that is that they use that 60 days to collect as MUCH data as possible on everyone using/trying it. With that, they are in a much better position business-wise.

that condition of it being free is more and more fading out

**Reply by moritztim** [reply]
> Opting out doesn't prevent copilot to [send all your data](https://www.reddit.com/r/programming/comments/vhnvlm/comment/idaad4b/?utm_source=share&utm_medium=web2x&context=3) tough
> 
> > If I understand the settings correctly then unchecking "Allow GitHub to use my code snippets for product improvements" should disable that.
> 
> > No, that's what allows them to use your code. No matter what, all the 'AI' stuff is done on a remote server that receives all your code. In most companies sending source code to a 3rd party without permission will get you fired almost instantly.

at least it is just using the code for the AI then and not for other things. You do still need permission for that tho.

**Comment by frozenprakash** [comment]
GitHub CoPilot (AI pair programmer) was only available on request basis since last year,

Now it's open for all the developers to give it a try and now it's available for VS IDE too along with VS Code.
Initially it'll be free for 60 days on trial basis, then it'll be billed.

Points to note:
1. Do not give any card info while signing up as you can simply skip that process and only try the trial
2. During installation it'll ask for usage stats, uncheck the checkbox which says share code for improving CoPilot

GitHub CoPilot Installation:
https://github.blog/2022-06-21-github-copilot-is-generally-available-to-all-developers/

This documentation should get you going right away:
https://github.com/github/copilot-docs

**Reply by bdollma** [reply]
How can you skip the billing process? All I get is:
<img width="706" alt="Screen Shot 2022-06-23 at 20 06 28" src="https://user-images.githubusercontent.com/43345270/175355327-91bb4f4a-a155-4dc1-b939-6f1c97c07820.png">

**Reply by frozenprakash** [reply]
Give Continue and don't provide the card details in the next step

**Reply by danny-does-stuff** [reply]
I'm looking at it right now and it doesn't look like there's an option to skip providing payment info

**Reply by ionics** [reply]
@danny-does-stuff Yes, seems this option of skipping providing payment information is gone.
